File: 649. Dota2 Senate/Solution.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r, d, rban, dban = 0, 0, 0, 0
        q = []
        for c in senate:
            q.append(c)
            if c == 'R':
                r += 1
            else:
                d += 1
        while r and d:
            curr = q.pop(0)
            if curr == 'D':
                if dban > 0:
                    dban -= 1
                    d -= 1
                else:
                    rban += 1
                    q.append(curr)
            else:
                if rban > 0:
                    rban -= 1
                    r -= 1
                else:
                    dban += 1
                    q.append(curr)
        return "Radiant" if r else "Dire"

File: 649. Dota2 Senate/Solution.java
class Solution {
    public String predictPartyVictory(String senate) {
        int r = 0, d = 0, rban = 0, dban = 0;
        Queue<Character> q = new LinkedList<>();
        for (char c : senate.toCharArray()) {
            q.offer(c);
            if (c == 'R') {
                r++;
            } else {
                d++;
            }
        }
        while (r > 0 && d > 0) {
            char curr = q.poll();
            if (curr == 'D') {
                if (dban > 0) {
                    dban--;
                    d--;
                } else {
                    rban++;
                    q.offer(curr);
                }
            } else {
                if (rban > 0) {
                    rban--;
                    r--;
                } else {
                    dban++;
                    q.offer(curr);
                }
            }
        }
        return r > 0 ? "Radiant" : "Dire";
    }
}